### DISCUSSION TITLE
Problems sorting by a relationship field

### DISCUSSION THREAD

**Discussion by dkanthofer** [payloadcms/payload#3978] [Feature Requests & Ideas]
### Link to reproduction

N/A

### Describe the Bug

When sorting a list by a relationship field, it actually sorts by the related document's ID. It should be alphabetically sorted by the field defined in `useAsTitle` of the related collection.

As you can see in the screenshot, the Testrelation column seems to be sorted randomly, while in fact this is the alphabetical order of the IDs from the related collection.

<img width="527" alt="image" src="https://github.com/payloadcms/payload/assets/3141586/4189614c-4a16-4378-aa9e-b316574c041e">

### To Reproduce

1. Add a field like this to a collection:
```
{
  name: 'testrelation',
  type: 'relationship',
  relationTo: 'relationExamples',
},
```

2. Then in the `relationExamples` collection, add at least a title field and define it as the useAsTitle:
```
const RelationExamples: CollectionConfig = {
  slug: 'relationExamples',
  admin: {
    useAsTitle: 'title',
    listSearchableFields: ['title'],
  },
  fields: [
    {
      name: 'title',
      type: 'text',
    },
  ],
}
```
3. Generate some entries in both collections and be sure to fill in the relationships
4. Now open the first collection and click one of the sort arrows next to the Testrelation column header.
5. You can check against the API that the sorting order is based on the ID, not the title

### Payload Version

2.0.13

### Adapters and Plugins

db-mongodb, bundler-webpack

**Comment by JessRynkar** [comment]
Hi @dkanthofer - currently this is the expected behavior. We would like to accomodate sorting on relationship fields in the future, which will require building some new functionality - so I'm going to convert this to a feature request and we'll get it on our roadmap.

~~For now, a potential workaround would be to customize the `id`. Adding a new text field with the name `id` to your collection will replace the default value. You could also use a field hook to automatically populate this field so you don't need to write it out twice.~~

EDIT: The workaround above is outdated and no longer recommended, it is likely to cause further issues and complications in your project.

**Reply by markusahrweileramcon** [reply]
is it also expected, that I cannot sort the collection by a field of the relationship?
e.g. createdAt?

`defaultSort: 'address.street'`
The only way to achieve this currently, would be to copy all fields of the relationships you possibly want to sort by to the entity itslef.
Which is heavy overload. When changing the field in the relationship, I need to update every entity, which has this relationship selected and update its field. Other way round, when I select a different relationship, I need to update the field also

**Reply by markusahrweileramcon** [reply]
is "virtual field" something different as i mentioned?

but thanks, for stating the obvious

**Reply by tlmader** [reply]
@BrianJM Sorry to bring up an old thread.

Payload's current docs on sorting state:

> Because sorting is handled by the database, the field cannot be a [Virtual Field](https://payloadcms.com/blog/learn-how-virtual-fields-can-help-solve-common-cms-challenges). It must be stored in the database to be searchable.

What's the current recommended approach to sorting on a related document's field?

**Reply by tamas-marton** [reply]
Hi @jessrynkar. Your suggested workaround sounds problematic for a number of reasons.
1. Even an overridden **ID field is immutable** (as it should be). If we populate the ID with some meaningful data (eg. title, name, company, etc.) on creation, it is doomed to live with that forever. There are many reasons to update such a value over time (typos, company name changes, etc.). Then you would have to:
   1. Create a duplicate of the related collection item with new ID.
   2. Update all relations to the new item.
   3. Delete the old related item.
I guess it all could work automagically with an extra field and hooks, but seems overkill.
2. The ID **should be unique**. It's easy to run into collisions with names and titles. We could use a visible-editable "source" field and generate IDs from it with an incremental numeric suffix generated, but the overhead makes me sad :)
3. There's a reasonable **limitation on text-type ID field values** (no dots and slashes). A person or company name can easily have those. You could work around it with a "source" field with those included and a sanitized ID generated from it (overhead again).

Am I getting your suggestion wrong?

Thankfully, **Payload is already capable of sorting the items by any (stored) field of related collections.** Adding a dot and the desired field name to the `sort` query parameter in the admin URL proves it. In OP's example, it's replacing `sort=-testrelation` with `sort=-testrelation.title`. So I guess, all it requires is a frontend patch. For a basic solution:
- the admin frontend could navigate to the `useTitle`-appended address
- and the sorting arrows needs to be aware of the change, when to stlye them active.

It could be useful to provide a dedicated sorting config option for `relationship` type fields (one might still want to sort by IDs, or a third field), but maybe it would be too complicated to implement.

I'm brand new to Payload, but might as well look into the code and send a PR for it. What do you guys think about the fix concept?

**Reply by markusahrweileramcon** [reply]
@tamas-marton are you sure, that the order is correct?
We use payload 3 and still have that problem.
We can see a change in the order, when we change the direction with a `-`, but it still orders by the **ID** of the related document and not by the provided field

**Reply by tamas-marton** [reply]
@markusahrweileramcon In case the user **manually rewrite the URL's `sort` query parameter** in the browser's address bar, it works indeed (in Payload v3). It **doesn't qualify even as a workaround**, since you can't expect users to rewrite URLs manually, plus the sorting arrow buttons all get an inactive style, so there's no visual clue about the current sorting. It's just to show that Payload's backend and admin UI **already handles such a sorting scenario by URL**, so it's probably just a matter of **patching some admin UI component(s)**.

**Reply by markusahrweileramcon** [reply]
@tamas-marton 
I directly used the api. Also, since v3 now uses SSR by default, we use the payload client and feed the parameters directly into the `find` function: `sort: searchParams.get('sort') || undefined`
![image](https://github.com/user-attachments/assets/b15a5b57-53d9-430e-a585-0cdcef1fb062)

**Reply by markusahrweileramcon** [reply]
Alright. After updating to 3.38.0 it looks like this feature is indeed implemented.
It's pretty unsatisfying, that this issue is not updated by staff

**Reply by tamas-marton** [reply]
> After updating to 3.38.0 it looks like this feature is indeed implemented.

Yeah, sorry for being vague with "v3", a lesson learned :)

**Reply by tamas-marton** [reply]
> It's pretty unsatisfying, that this issue is not updated by staff

Come on man. It's free software, explicitly stating "WITHOUT WARRANTY OF ANY KIND" in its license. I think it's fair to say that Payload developers owe nothing to us users (regarding either code improvements or discussion/issue management), but still work hard on resolving problems. There's no restriction on modifying your own instance, opening a PR or even start off a fork if you're unsatisfied with something.

**Comment by mathieu2em** [comment]
hi guys is it still the current behavior ? 
@JessChowdhury I tried your solution with the idea of customizing the id but problem is it breaks the relationship as the relationship is linked by its id to the collection

**Reply by denolfe** [reply]
Sorting by other values should work. [Field in our test suite sorted by text here](https://github.com/payloadcms/payload/blob/main/test/fields/collections/Relationship/index.ts#L30). If you're seeing otherwise, I'd log an issue.

**Reply by mathieu2em** [reply]
Hi @denolfe thank you for your answer,
I believe you are talking about the behavior in the dropdown while we are talking about the behavior in the admin panel when looking at the collection.
I updated to the latest version and tried what you just showed in your link but it's not changing the behavior when I sort in my admin collection panel.
I would like the behavior to be that when I click the little arrows, on my field to sort it, for it to sort by the sortOptions that I stated or at least by what I put in the collection in useAsTitle or defaultSort.

**Reply by mathieu2em** [reply]
<img width="1386" alt="Screenshot 2024-02-28 at 4 14 25 PM" src="https://github.com/payloadcms/payload/assets/35858630/6e50ac48-c974-4ec9-a738-14e68e7a18e2">

**Comment by matthijs166** [comment]
Is this supported in Payload 3?

**Reply by markusahrweileramcon** [reply]
afaik, no

**Reply by matthijs166** [reply]
For all our next project we are going to use the search plugin if we have overviews where we need sorting on deeper levels.

**Comment by JessRynkar** [comment]
Hi all – thanks for all the input here and sorry for the delayed response on this thread.

To clarify: 
- by default, sorting the collection list by a specific field within a relationship is still not supported as of Payload `v3.38.0` 
- when sorting by a relationship field in the admin UI, the sort query uses the related document ID, not the `useAsTitle` or any other field

We agree this behavior should be accommodated - ideally by sorting on `useAsTitle` or via a configurable option. I have added this as a new feature request so we can properly track and prioritize this on our roadmap [here](https://github.com/payloadcms/payload/discussions/12451).

Currently our recommended workaround is to use virtual fields as outlined in [this recent feature](https://github.com/payloadcms/payload/pull/11805). For example, if I have a `posts` collection with a relationship to an `author` collection, I can implement a virtual field for `author.name` by doing the following:
```ts
collections: [
    {
      slug: 'posts',
      fields: [
        {
          name: 'title',
          type: 'text',
        },
        {
          name: 'author',
          type: 'relationship',
          relationTo: 'author',
        },
        // To sort by the author name, instead of the ID, assign it to a virtual field
        // This will create a new column in the collection view that can be used to sort
        {
          name: 'authorName',
          virtual: 'author.name',
          type: 'text',
        },
      ],
    },
    {
      slug: 'author',
      fields: [
        {
          name: 'name',
          type: 'text',
        },
      ],
    },
  ],
```

This virtual field will be displayed in the collection list view and can be used to sort as expected.

**Reply by tamas-marton** [reply]
Yay! Thanks for pointing out the recent feature of improved virtual field too.
